functions.cmake

```diff
@@ -232,7 +233,13 @@ function(_add_library)
     list(FIND CLR_CROSS_COMPONENTS_LIST ${ARGV0} INDEX)  
     if (DEFINED CLR_CROSS_COMPONENTS_LIST AND ${INDEX} EQUAL -1)  
      set_target_properties(${ARGV0} PROPERTIES EXCLUDE_FROM_ALL 1)  
-    endif()  
+    endif()
+    set_property(GLOBAL APPEND_STRING PROPERTY CLR_CMAKE_CORECLR_TARGETS "${ARGV0};")  
+endfunction()
+
+function(_add_custom_target)
```

Comment:
Why do we still need to cmake changes once you switched to Directory.Build.props?

Reply:
There are 3 reasons I have decided it is worth to keep it:

1. Centralizing CMake logic for target addition (library, executable, custom target) allows to detect build logic bugs which otherwise would be missed - see comment https://github.com/dotnet/coreclr/pull/14578#issuecomment-337929416 on unneeded superpmi assemblies being added to cross components binaries bug which is tracked by issue #14596.

2. Allows to continue supporting VS 2015 (MSBuild 14) based builds with current CMake based method (VS_GLOBAL_CL_MPCount property).

3. Allows to prevent picking up by MSBuild extension point any other `Directory.Build.props` file by limiting overriding of `DirectoryBuildPropsPath` with auto discovered files in directory tree below project root. It is achieved by injecting `DirectoryBuildPropsPath` property set to absolute path to `$(RepoRoot)/Directory.Build.props` using VS_GLOBAL_DirectoryBuildPropsPath property.

Reply:
I plan to write short supplement to native build dcos to describe Hot To for current CMake logic and best practises when adding new project/files.

Reply:
I was hoping that we will use the BUILDSYSTEM_TARGETS and require cmake >= 3.7 instead of manually creating the property and having to modify the _add_xxxxxx functions. Is there aomething preventing us doing that?

Reply:
For (1) I agree with @janvorli if we can avoid needing to override that method then we should try to do that.

For (2) I don't feel it is necessary as VS2015 support will be removed as soon as we get all our official build and CI machines with VS2017 installed so I wouldn't worry about that reason. 

For (3) why do you feel you need to override the DirectoryBuildPropsPath? That is essentially killing the extensibility hook for anyone else that may try to use it for other purposes in other layers of the directory structure, I would avoid that if we can.